File: leetcode/python/0003_longest_substring_without_repeating_characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/
# O(n) time complexity
# O(min(n,m)) space complexity
# n is the length of input string
# m is the size of the charset
def length_of_longest_substring(s):
    left_bound = max_length = 0
    dictionary = {}
    for right_bound, char in enumerate(s):
        if char in dictionary:
            left_bound = max(left_bound, dictionary[char] + 1)
        dictionary[char] = right_bound
        max_length = max(max_length, right_bound - left_bound + 1)
    return max_length

File: leetcode/java/0003_longest_substring_without_repeating_characters.java
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// O(n) time complexity
// O(min(m, n)) space complexity, where m is the size of the charSet.
import java.util.*;

class Solution_0003 {
  public int lengthOfLongestSubstring(String s) {
    Map<Character, Integer> map = new HashMap<>();
    int maxLength = 0;

    for (int leftIdx = 0, rightIdx = 0; rightIdx < s.length(); rightIdx++) {
      char rightChar = s.charAt(rightIdx);
      if (map.containsKey(rightChar)) {
        leftIdx = Math.max(map.get(rightChar) + 1, leftIdx);
      }
      map.put(rightChar, rightIdx);
      maxLength = Math.max(rightIdx - leftIdx + 1, maxLength);
    }

    return maxLength;
  }
}